lib/block-supports/colors.php

```diff
@@ -76,66 +76,41 @@ function gutenberg_apply_colors_support( $block_type, $block_attributes ) {
+		$custom_text_color          = _wp_array_get( $block_attributes, array( 'style', 'color', 'text' ), null );
+		$color_block_styles['text'] = $preset_text_color ? $preset_text_color : $custom_text_color;
 	}
 
 	// Background colors.
 	if ( $has_background_colors_support && ! gutenberg_should_skip_block_supports_serialization( $block_type, 'color', 'background' ) ) {
-		$has_named_background_color  = array_key_exists( 'backgroundColor', $block_attributes );
-		$has_custom_background_color = isset( $block_attributes['style']['color']['background'] );
-
-		// Apply required background class.
-		if ( $has_custom_background_color || $has_named_background_color ) {
-			$classes[] = 'has-background';
-		}
-		// Apply background color classes or styles.
-		if ( $has_named_background_color ) {
-			$classes[] = sprintf( 'has-%s-background-color', _wp_to_kebab_case( $block_attributes['backgroundColor'] ) );
-		} elseif ( $has_custom_background_color ) {
-			$styles[] = sprintf( 'background-color: %s;', $block_attributes['style']['color']['background'] );
-		}
+		$preset_background_color          = array_key_exists( 'backgroundColor', $block_attributes ) ? "var:preset|background-color|{$block_attributes['backgroundColor']}" : null;
+		$custom_background_color          = _wp_array_get( $block_attributes, array( 'style', 'color', 'background' ), null );
+		$color_block_styles['background'] = $preset_background_color ? $preset_background_color : $custom_background_color;
 	}
 
 	// Gradients.
+
 	if ( $has_gradients_support && ! gutenberg_should_skip_block_supports_serialization( $block_type, 'color', 'gradients' ) ) {
-		$has_named_gradient  = array_key_exists( 'gradient', $block_attributes );
-		$has_custom_gradient = isset( $block_attributes['style']['color']['gradient'] );
-
-		if ( $has_named_gradient || $has_custom_gradient ) {
-			$classes[] = 'has-background';
-		}
-		// Apply required background class.
-		if ( $has_named_gradient ) {
-			$classes[] = sprintf( 'has-%s-gradient-background', _wp_to_kebab_case( $block_attributes['gradient'] ) );
-		} elseif ( $has_custom_gradient ) {
-			$styles[] = sprintf( 'background: %s;', $block_attributes['style']['color']['gradient'] );
-		}
+		$preset_gradient_color          = array_key_exists( 'gradient', $block_attributes ) ? "var:preset|background|{$block_attributes['gradient']}" : null;
+		$custom_gradient_color          = _wp_array_get( $block_attributes, array( 'style', 'color', 'gradient' ), null );
+		$color_block_styles['gradient'] = $preset_gradient_color ? $preset_gradient_color : $custom_gradient_color;
 	}
 
-	$attributes = array();
-	if ( ! empty( $classes ) ) {
-		$attributes['class'] = implode( ' ', $classes );
+	$attributes   = array();
+	$style_engine = gutenberg_get_style_engine();
+	$styles       = $style_engine->generate( array( 'color' => $color_block_styles ) );
```

Comment:
From the API point of view, I guess we could do
```
$styles       = gutenberg_style_engine_generate( array( 'color' => $color_block_styles ) );
or 
$styles       = gutenberg_style_engine_get_styles( array( 'color' => $color_block_styles ) );
```
And avoid the need to call gutenberg_get_style_engine. That call may be done inside gutenberg_style_engine_generate.

Not directly related to this PR but regarding the class WP_Style_Engine. It seems it will be part of WordPress 6.0 and it's a public class everyone can use and we need to maintain forever. Are we totally confident this class is ready and useful for public usage?
To me this class seems like a helper to a function that generates styles like gutenberg_style_engine_generate or gutenberg_style_engine_get_styles. I guess we could mark for now the class  WP_Style_Engine as private with something similar to what we did to the theme.json class:
```
 * This class is for internal core usage and is not supposed to be used by extenders (plugins and/or themes).
 * This is a low-level API that may need to do breaking changes. Please,
 * use gutenberg_style_engine_get_styles instead.
 *
 * @access private
```

Reply:
> And avoid the need to call gutenberg_get_style_engine. That call may be done inside gutenberg_style_engine_generate.

That's a fair point. I'll look into it 👍 The easier it is to call the better.

> It seems it will be part of WordPress 6.0 and it's a public class everyone can use and we need to maintain forever. Are we totally confident this class is ready and useful for public usage?

Good question.

We decided [**not** to port it over for WordPress 6.0](https://github.com/WordPress/gutenberg/issues/39889#issuecomment-1086985048), but it's relevant for 6.1. 

The [JS version of the style engine](https://github.com/WordPress/gutenberg/pull/37978) has been in Gutenberg for some time now, and the backend is trying to mirror its interface and functionality as best as it can.

Theoretically, the only public-facing API method should be `gutenberg_style_engine_generate`. That method – its arguments and return value – shouldn't need to change. 

However innards of the class will be in flux for some time, so I guess the honest answer is "not quite".

> To me this class seems like a helper to a function that generates styles like gutenberg_style_engine_generate or gutenberg_style_engine_get_styles.

Yeah, that's a fair way of looking at it. Ultimately, the style engine should be a black box while we're still working on its functionality.

It's tough because we want to iterate, but we also want to make it clear that things are still experimental.

>  I guess we could mark for now the class WP_Style_Engine as private with something similar to what we did to the theme.json

Great idea, thanks. 

We don't really want to expose the implementation details, so it's definitely something we should communicate here.